### ISSUE TITLE
Build fails with version 2.0.3 and 2.0.2, works with v2.0.1

### ISSUE THREAD

**Issue by DanielSWolf** [helaili/jekyll-action#45]
I created a simple little Jekyll project using `jekyll-action@2.0.3`. When the GitHub build runs, the action fails with the message _Cannot publish on branch master_. I don't understand where this is coming from, since the action ought to push to the `gh-pages` branch and I didn't change any of its configuration.

I get the same error with version 2.0.2. Version 2.0.1 works fine.

Here is [my repo](https://github.com/DanielSWolf/DanielSWolf.github.io).

Here is the full log:

```
2020-07-04T16:31:46.9030154Z     Use helaili/jekyll-action@2.0.3
2020-07-04T16:31:46.9119585Z ##[group]Run helaili/jekyll-action@2.0.3
2020-07-04T16:31:46.9119739Z with:
2020-07-04T16:31:46.9119865Z env:
2020-07-04T16:31:46.9120443Z   JEKYLL_PAT: ***
2020-07-04T16:31:46.9120571Z ##[endgroup]
2020-07-04T16:31:46.9143811Z ##[command]/usr/bin/docker run --name d31be7f02408bd4b339a04464a8d228e90_90cfa8 --label 3888d3 --workdir /github/workspace --rm -e JEKYLL_PAT -e INPUT_JEKYLL_SRC -e HOME -e GITHUB_JOB -e GITHUB_REF -e GITHUB_SHA -e GITHUB_REPOSITORY -e GITHUB_REPOSITORY_OWNER -e GITHUB_RUN_ID -e GITHUB_RUN_NUMBER -e GITHUB_ACTOR -e GITHUB_WORKFLOW -e GITHUB_HEAD_REF -e GITHUB_BASE_REF -e GITHUB_EVENT_NAME -e GITHUB_SERVER_URL -e GITHUB_API_URL -e GITHUB_GRAPHQL_URL -e GITHUB_WORKSPACE -e GITHUB_ACTION -e GITHUB_EVENT_PATH -e RUNNER_OS -e RUNNER_TOOL_CACHE -e RUNNER_TEMP -e RUNNER_WORKSPACE -e ACTIONS_RUNTIME_URL -e ACTIONS_RUNTIME_TOKEN -e ACTIONS_CACHE_URL -e GITHUB_ACTIONS=true -e CI=true -v "/var/run/docker.sock":"/var/run/docker.sock" -v "/home/runner/work/_temp/_github_home":"/github/home" -v "/home/runner/work/_temp/_github_workflow":"/github/workflow" -v "/home/runner/work/DanielSWolf.github.io/DanielSWolf.github.io":"/github/workspace" 3888d3:1be7f02408bd4b339a04464a8d228e90  ""
2020-07-04T16:31:47.2549604Z Starting the Jekyll Action
2020-07-04T16:31:49.4134948Z Fetching gem metadata from https://rubygems.org/..........
2020-07-04T16:31:49.4868419Z Fetching gem metadata from https://rubygems.org/.
2020-07-04T16:31:49.5247881Z Resolving dependencies...
2020-07-04T16:31:49.5443305Z Fetching public_suffix 4.0.5
2020-07-04T16:31:49.6039769Z Installing public_suffix 4.0.5
2020-07-04T16:31:49.6251807Z Fetching addressable 2.7.0
2020-07-04T16:31:49.6828318Z Installing addressable 2.7.0
2020-07-04T16:31:49.6986989Z Fetching asciidoctor 2.0.10
2020-07-04T16:31:49.7726819Z Installing asciidoctor 2.0.10
2020-07-04T16:31:49.8203725Z Using bundler 2.1.4
2020-07-04T16:31:49.8203974Z Fetching colorator 1.1.0
2020-07-04T16:31:49.8602022Z Installing colorator 1.1.0
2020-07-04T16:31:49.8745253Z Fetching concurrent-ruby 1.1.6
2020-07-04T16:31:49.9487571Z Installing concurrent-ruby 1.1.6
2020-07-04T16:31:50.0124910Z Fetching eventmachine 1.2.7
2020-07-04T16:31:50.0827915Z Installing eventmachine 1.2.7 with native extensions
2020-07-04T16:31:59.7901710Z Fetching http_parser.rb 0.6.0
2020-07-04T16:31:59.8899932Z Installing http_parser.rb 0.6.0 with native extensions
2020-07-04T16:32:01.0846021Z Fetching em-websocket 0.5.1
2020-07-04T16:32:01.1440581Z Installing em-websocket 0.5.1
2020-07-04T16:32:01.1662499Z Fetching ffi 1.13.1
2020-07-04T16:32:01.2797423Z Installing ffi 1.13.1 with native extensions
2020-07-04T16:32:09.5567424Z Fetching forwardable-extended 2.6.0
2020-07-04T16:32:09.6049883Z Installing forwardable-extended 2.6.0
2020-07-04T16:32:09.6137419Z Fetching i18n 1.8.3
2020-07-04T16:32:09.6698408Z Installing i18n 1.8.3
2020-07-04T16:32:09.7051026Z Fetching sassc 2.4.0
2020-07-04T16:32:09.7883532Z Installing sassc 2.4.0 with native extensions
2020-07-04T16:33:33.1957679Z Fetching jekyll-sass-converter 2.1.0
2020-07-04T16:33:33.2479714Z Installing jekyll-sass-converter 2.1.0
2020-07-04T16:33:33.2561978Z Fetching rb-fsevent 0.10.4
2020-07-04T16:33:33.3180655Z Installing rb-fsevent 0.10.4
2020-07-04T16:33:33.3378009Z Fetching rb-inotify 0.10.1
2020-07-04T16:33:33.3927433Z Installing rb-inotify 0.10.1
2020-07-04T16:33:33.4068527Z Fetching listen 3.2.1
2020-07-04T16:33:33.4586714Z Installing listen 3.2.1
2020-07-04T16:33:33.4782839Z Fetching jekyll-watch 2.2.1
2020-07-04T16:33:33.5188535Z Installing jekyll-watch 2.2.1
2020-07-04T16:33:33.5273178Z Fetching rexml 3.2.4
2020-07-04T16:33:33.5816261Z Installing rexml 3.2.4
2020-07-04T16:33:33.6180249Z Fetching kramdown 2.3.0
2020-07-04T16:33:33.6884333Z Installing kramdown 2.3.0
2020-07-04T16:33:33.8951562Z Fetching kramdown-parser-gfm 1.1.0
2020-07-04T16:33:33.9385819Z Installing kramdown-parser-gfm 1.1.0
2020-07-04T16:33:33.9555808Z Fetching liquid 4.0.3
2020-07-04T16:33:34.0182375Z Installing liquid 4.0.3
2020-07-04T16:33:34.0633471Z Fetching mercenary 0.4.0
2020-07-04T16:33:34.1114901Z Installing mercenary 0.4.0
2020-07-04T16:33:34.1248042Z Fetching pathutil 0.16.2
2020-07-04T16:33:34.1657472Z Installing pathutil 0.16.2
2020-07-04T16:33:34.1728346Z Fetching rouge 3.20.0
2020-07-04T16:33:34.3283084Z Installing rouge 3.20.0
2020-07-04T16:33:34.5044059Z Fetching safe_yaml 1.0.5
2020-07-04T16:33:34.5526717Z Installing safe_yaml 1.0.5
2020-07-04T16:33:34.5741506Z Fetching unicode-display_width 1.7.0
2020-07-04T16:33:34.6136700Z Installing unicode-display_width 1.7.0
2020-07-04T16:33:34.6207849Z Fetching terminal-table 1.8.0
2020-07-04T16:33:34.6658710Z Installing terminal-table 1.8.0
2020-07-04T16:33:34.6760018Z Fetching jekyll 4.1.1
2020-07-04T16:33:34.7435096Z Installing jekyll 4.1.1
2020-07-04T16:33:34.7951819Z Fetching jekyll-asciidoc 3.0.0
2020-07-04T16:33:34.8443609Z Installing jekyll-asciidoc 3.0.0
2020-07-04T16:33:34.8581473Z Fetching jekyll-feed 0.14.0
2020-07-04T16:33:34.9051023Z Installing jekyll-feed 0.14.0
2020-07-04T16:33:34.9133254Z Fetching jekyll-seo-tag 2.6.1
2020-07-04T16:33:34.9651574Z Installing jekyll-seo-tag 2.6.1
2020-07-04T16:33:34.9799753Z Fetching minima 2.5.1
2020-07-04T16:33:35.0268971Z Installing minima 2.5.1
2020-07-04T16:33:35.0408856Z Bundle complete! 7 Gemfile dependencies, 33 gems now installed.
2020-07-04T16:33:35.0409089Z Bundled gems are installed into `./vendor/bundle`
2020-07-04T16:33:35.0409929Z Post-install message from i18n:
2020-07-04T16:33:35.0410082Z 
2020-07-04T16:33:35.0410389Z HEADS UP! i18n 1.1 changed fallbacks to exclude default locale.
2020-07-04T16:33:35.0410567Z But that may break your application.
2020-07-04T16:33:35.0410651Z 
2020-07-04T16:33:35.0410841Z If you are upgrading your Rails application from an older version of Rails:
2020-07-04T16:33:35.0410959Z 
2020-07-04T16:33:35.0411357Z Please check your Rails app for 'config.i18n.fallbacks = true'.
2020-07-04T16:33:35.0411768Z If you're using I18n (>= 1.1.0) and Rails (< 5.2.2), this should be
2020-07-04T16:33:35.0412159Z 'config.i18n.fallbacks = [I18n.default_locale]'.
2020-07-04T16:33:35.0412338Z If not, fallbacks will be broken in your app by I18n 1.1.x.
2020-07-04T16:33:35.0412446Z 
2020-07-04T16:33:35.0412611Z If you are starting a NEW Rails application, you can ignore this notice.
2020-07-04T16:33:35.0412711Z 
2020-07-04T16:33:35.0412854Z For more info see:
2020-07-04T16:33:35.0413413Z https://github.com/svenfuchs/i18n/releases/tag/v1.1.0
2020-07-04T16:33:35.0413571Z 
2020-07-04T16:33:35.5844462Z Configuration file: /github/workspace/_config.yml
2020-07-04T16:33:35.6348557Z             Source: /github/workspace
2020-07-04T16:33:35.6348794Z        Destination: /github/workspace/build
2020-07-04T16:33:35.6349688Z  Incremental build: disabled. Enable with --incremental
2020-07-04T16:33:35.6349868Z       Generating... 
2020-07-04T16:33:35.6655914Z        Jekyll Feed: Generating feed for posts
2020-07-04T16:33:35.7448430Z                     done in 0.11 seconds.
2020-07-04T16:33:35.7449375Z  Auto-regeneration: disabled. Use --watch to enable.
2020-07-04T16:33:35.7480884Z Jekyll build done
2020-07-04T16:33:35.7486099Z Cannot publish on branch master
```

**Comment by DanielSWolf** [comment]
Is there any additional information I can provide to narrow down the problem?

**Comment by helaili** [comment]
Can you create a `ACTIONS_STEP_DEBUG` secret with value `true` so we can get more info?

**Comment by bertolinocastro** [comment]
Hi! I probably I'm having the same issue as @DanielSWolf so I'll post both Actions' logs here (v2.0.3 & v2.0.1). Like what happend to him, my build also only got to work when I downgraded the Action version to v2.0.1. I used this `ACTIONS_STEP_DEBUG` as well.

[logs_2.0.3.zip](https://github.com/helaili/jekyll-action/files/4947024/logs_7.zip)  

[logs_2.0.1.zip](https://github.com/helaili/jekyll-action/files/4947031/logs_8.zip)

**Comment by bertolinocastro** [comment]
While reading the code, it looked like that this error appears only when pushing from the master branch in user.github.io repositories. Changing the name of the branch & pushing from it solved this error.

**Comment by DanielSWolf** [comment]
For simple Git repos without Gitflow, `master` is the default branch. So I'm not keen on renaming my default branch to work around this bug. Can this be fixed?

**Comment by DanielSWolf** [comment]
Now I get it! This plugin uses a heuristic to determine the target branch:

* If the repository ends with `.github.io`, it pushes to the `master` branch.
* Otherwise, it pushes to the `gh-pages` branch.

If the target branch is identical with the source branch, the plugin aborts with the message "::error::Cannot publish on branch <target branch>".

I didn't realize that GitHub used different branch naming conventions for "regular" repos and `.github.io` repos.

On re-reading your README file, I realized that it actually mentions this behavior, though I didn't understand it initially. It might be helpful to go into a bit more detail in the README file so that others won't stumble here.

**Comment by DanielSWolf** [comment]
Also, it would be great if I could explicitly configure the target branch, overriding the default heuristic!

**Comment by bertolinocastro** [comment]
@DanielSWolf this heuristic is somehow precise regarding the protocol GitHub uses to offer a static page for repositories. If you push the static files of your site in any branch except the `master` while this is a `"user".github.io" repo, it won't be published. For this kind of repo, GitHub requires the static files to be in the `master` branch.

Otherwise, for non-user repo, GitHub uses the `gh-pages` branch by default (although they allow us to change which branch to publish, if I'm not wrong). Maybe in this case the `jekyll-action` code should have an ENV var to allow the changing of branch to push, if it doesn't already have it.

**Comment by helaili** [comment]
Fixed by #51
